### PR TITLE
fix: replace type assertion with type guard

### PR DESCRIPTION
**Pull request by dnalborczyk**
This PR contains:

- [ ] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [x] other

Are tests included?

- [ ] yes (_bugfixes and features will not be merged without tests_)
- [x] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

### CODE REVIEW

File: src/Bundle.ts [hunk 1]
```diff
@@ -306,11 +306,11 @@ function validateOptionsForMultiChunkOutput(
 }
 
 function getIncludedModules(modulesById: Map<string, Module | ExternalModule>): Module[] {
-	return [...modulesById.values()].filter(
-		module =>
+	return Array.from(modulesById.values()).filter(
```
Comment (lukastaegert) [inline src/Bundle.ts:309]: Out of curiosity: Does `Array.from` provide any typing benefits over `[... iterator]`? From their behaviour if you do not use the optional second argument of `Array.from`, I would expect them to be more or less the same with the spread operator possibly being slightly more efficient as it needs to do less checking.

File: src/utils/getOriginalLocation.ts [hunk 2]
```diff
@@ -4,10 +4,9 @@ export function getOriginalLocation(
 	sourcemapChain: DecodedSourceMapOrMissing[],
 	location: { column: number; line: number; name?: string; source?: string }
 ): { column: number; line: number } {
-	// This cast is guaranteed. If it were a missing Map, it wouldn't have a mappings.
 	const filteredSourcemapChain = sourcemapChain.filter(
-		sourcemap => sourcemap.mappings
-	) as ExistingDecodedSourceMap[];
+		(sourcemap): sourcemap is ExistingDecodedSourceMap => sourcemap.mappings !== undefined
```
Comment (lukastaegert) [inline src/utils/getOriginalLocation.ts:8]: Are you sure you want to explicitly compare against undefined and there will never be a `null`? The "good" type should be an object, so it would be ok to also filter out e.g. empty strings etc. I think I liked the previous check better.

**Review by lukastaegert — APPROVED** [review]

File: src/Bundle.ts (hunk 1 above)
Comment (dnalborczyk) [reply src/Bundle.ts]: I think it's more or less the same. dunno about the runtime tho, as the former creates a new array, but all that might be optimized away anyways. 🤷

File: src/utils/getOriginalLocation.ts (hunk 2 above)
Comment (dnalborczyk) [reply src/utils/getOriginalLocation.ts]: @lukastaegert   uh, sorry, I think I came too late. :/

compare the original source code: `sourcemap => sourcemap.mappings` ---> no comparison at all

I addd it only because according to the types, `mappings` can be an `array` and `undefined`.

```
export type DecodedSourceMapOrMissing =
	| {
			mappings?: never;
			missing: true;
			plugin: string;
	  }
	| ExistingDecodedSourceMap;
```

> I think I liked the previous check better.

there were none. ;)

Comment (lukastaegert) [reply src/utils/getOriginalLocation.ts]: The previous check was a check for truthiness, which you replaced with an explicit but much more strict one.

Comment (dnalborczyk) [reply src/utils/getOriginalLocation.ts]: the previous check seemed to be insufficient for typescript, otherwise I would have left it as is. 😉 

https://github.com/rollup/rollup/runs/4639006105?check_suite_focus=true

I assume that `filter` expects a `real boolean`, not just a `truthy/falsy` value. (at least in the world of TS).

do you think the typing is incorrect then? can `mappings` be null as well? presumably a check for `!= null` would be better? to check for `falsy` or `thruthy` values seems to be over checking for things. if mappings can only be an `array` or `undefined` (and possibly `null`) I don't think it's good pattern to check for anything else truthy (0, '', and what not) _just in case_, unless that case is valid. in that case tho, the typings might be incorrect.

I think it makes the code more readable and more robust. it's really hard to follow any code paths otherwise, specially if the code and the types don't match.

Comment (lukastaegert) [reply src/utils/getOriginalLocation.ts]: I vaguely remember that mappings could be an empty string, but I am not sure in this instance. Moreover, anything could be in a user generated source map. I am just saying, the previous check was concise and worked well, also for TypeScript. TypeScript understands very well that filter does an implicit cast to Boolean and does not flag anything here. The new check might work as well, but it needlessly takes the risk of breaking stuff for other people, and I do not see a need to do so here.
